### PR TITLE
docs(asgi-tutorial): include info on setting up logging for debugging

### PR DESCRIPTION
**Pull request by MRLab12**
# Summary of Changes

This PR introduces a new section to the ASGI tutorial in the Falcon documentation, aimed at providing developers with clear instructions on setting up and utilizing logging for debugging ASGI applications. The added content outlines the steps to configure basic logging using Python's built-in logging module, offering examples of how to log informational, warning, and error messages effectively.

# Related Issues

#2098 

# Pull Request Checklist

This is just a reminder about the most common mistakes.  Please make sure that you tick all *appropriate* boxes.  But please read our [contribution guide](../CONTRIBUTING.md) at least once; it will save you a few review cycles!

If an item doesn't apply to your pull request, **check it anyway** to make it apparent that there's nothing to do.

- [ ] Applied changes to both WSGI and ASGI code paths and interfaces (where applicable).
- [ ] Added **tests** for changed code.
- [ ] Prefixed code comments with GitHub nick and an appropriate prefix.
- [ ] Coding style is consistent with the rest of the framework.
- [ ] Updated **documentation** for changed code.
    - [ ] Added docstrings for any new classes, functions, or modules.
    - [ ] Updated docstrings for any modifications to existing code.
    - [ ] Updated both WSGI and ASGI docs (where applicable).
    - [ ] Added references to new classes, functions, or modules to the relevant RST file under `docs/`.
    - [ ] Updated all relevant supporting documentation files under `docs/`.
    - [ ] A copyright notice is included at the top of any new modules (using your own name or the name of your organization).
    - [ ] Changed/added classes/methods/functions have appropriate `versionadded`, `versionchanged`, or `deprecated` [directives](http://www.sphinx-doc.org/en/stable/usage/restructuredtext/directives.html?highlight=versionadded#directive-versionadded).
- [ ] Changes (and possible deprecations) have [towncrier](https://towncrier.readthedocs.io/en/latest/quickstart.html#creating-news-fragments) news fragments under `docs/_newsfragments/`, with the file name format `{issue_number}.{fragment_type}.rst`. (Run `towncrier --draft` to ensure it renders correctly.)

If you have *any* questions to *any* of the points above, just **submit and ask**! This checklist is here to *help* you, not to deter you from contributing!

*PR template inspired by the attrs project.*

### CODE REVIEW

File: docs/user/tutorial-asgi.rst [hunk 1]
```diff
@@ -967,6 +967,47 @@ adding ``--cov-fail-under=100`` (or any other percent threshold) to our
...
+
+.. code:: python
+
+    try:
+        # Some operation...
+    except Exception as e:
+        logger.exception('An error occurred: %s', str(e))
+
+
+For more sophisticated logging setups (e.g., different log levels or formats for development and production), you can configure multiple handlers and formatters, as described in the Python logging documentation.
```
Comment (CaselIT) [inline docs/user/tutorial-asgi.rst:1006]: we may want to link to the external logging docs: https://docs.python.org/3/howto/logging.html#logging-basic-tutorial

File: docs/user/tutorial-asgi.rst [hunk 2]
```diff
@@ -967,6 +967,47 @@ adding ``--cov-fail-under=100`` (or any other percent threshold) to our
     tests in multiple environments would most probably involve running
     ``coverage`` directly, and combining results.
 
+Debugging ASGI Applications
+---------------------------
+
+While developing and testing ASGI applications, understanding how to configure and utilize logging can be helpful, especially when you encounter unexpected issues or behaviors.
```
Comment (CaselIT) [inline docs/user/tutorial-asgi.rst:973]: We could add "This section applies also when using WSGI application"

**Review by CaselIT — CHANGES_REQUESTED** [review]
I think the lines should in the rst go to a new line at about 80-100 chars.

Other than that it seems a good addition, thanks!

File: docs/user/tutorial-asgi.rst (hunk 1 above)
Comment (MRLab12) [reply docs/user/tutorial-asgi.rst:1006]: Yup, good idea.

File: docs/user/tutorial-asgi.rst [hunk 3]
```diff
@@ -967,6 +967,49 @@ adding ``--cov-fail-under=100`` (or any other percent threshold) to our
     tests in multiple environments would most probably involve running
     ``coverage`` directly, and combining results.
 
+Debugging ASGI Applications
+---------------------------
+(This section also applies to WSGI applications)
+
+While developing and testing ASGI applications, understanding how to configure and utilize logging can be helpful, especially when you encounter unexpected issues or behaviors.
+
+By default, Falcon does not set up logging for you, but Python's built-in `logging` module provides a flexible framework for emitting and capturing log messages. Here's how you can set up basic logging in your ASGI Falcon application:
```
Comment (vytas7) [inline docs/user/tutorial-asgi.rst:976]: IIRC, we have intersphinx connected to the stdlib's docs, so you probably can even link to them:
```
:mod:`logging`
```

File: docs/user/tutorial-asgi.rst [hunk 4]
```diff
@@ -967,6 +967,49 @@ adding ``--cov-fail-under=100`` (or any other percent threshold) to our
...
+
+You can now use `logger` to log messages within your application. For example:
+
+.. code:: python
+
+    logger.info('Info message')
+    logger.warning('Warning message')
+    logger.error('Error message')
+
+It's especially useful to log exceptions and error messages that can help diagnose issues:
```
Comment (vytas7) [inline docs/user/tutorial-asgi.rst:997]: This is a good intro on logging, but IMHO the reader of this second tutorial is already expected to be an experienced Python developer, so maybe we could shorten this generic section a bit, and tell more about Falcon's default `Exception` handler?

In addition to rendering an HTTP 500 response, it does log a message to the `falcon` logger. Maybe we could even show how this looks like?

**Review by vytas7 — CHANGES_REQUESTED** [review]
Thanks for this improvement, it looks good for the most of it!

I've left some suggestions inline, and as pointed out by @CaselIT, we should shorten RST lines to a reasonable length around 80 characters.

File: docs/user/tutorial-asgi.rst (hunk 3 above)
Comment (MRLab12) [reply docs/user/tutorial-asgi.rst:976]: Ah had not seen there was an intersphinx mapping. Doing this change.

File: docs/user/tutorial-asgi.rst (hunk 4 above)
Comment (MRLab12) [reply docs/user/tutorial-asgi.rst:997]: Yeah totally makes sense. Is something like this what you have in mind?

```
import falcon
import logging

logging.basicConfig(level=logging.INFO)

class ErrorResource:
    def on_get(self, req, resp):
        raise Exception("Something went wrong!")

app = falcon.App()
app.add_route('/error', ErrorResource())
```

File: docs/user/tutorial-asgi.rst [hunk 5]
```diff
@@ -967,6 +967,54 @@ adding ``--cov-fail-under=100`` (or any other percent threshold) to our
...
+.. code:: python
+
+    import falcon
+    import logging
+
+    logging.basicConfig(level=logging.INFO)
+
+    class ErrorResource:
+        def on_get(self, req, resp):
+            raise Exception("Something went wrong!")
```
Comment (vytas7) [inline docs/user/tutorial-asgi.rst:992]: I think we are using single quotes throughout the codebase

**Review by vytas7 — APPROVED** [review]
I took the liberty of changing your `Exception` line to use single quotes, looks good otherwise now, thanks for this improvement!

Another thought is that we probably want to move this section somewhere closer to the start, because it is likely that the reader might encounter errors earlier :slightly_smiling_face: But we can follow up on this later. LGTM _as is_ for now.

**Comment by vytas7** [comment]
Oh, it seems that we have some CI fighting to do that is unrelated to the proposed changes :grimacing: I'll take a look at these tomorrow.

**Review by CaselIT — APPROVED** [review]
looks good to me

**Pull request merged**